src/csharp/NativeMethods.cs

```diff
@@ -0,0 +1,90 @@
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaCreateGeneratorParams(IntPtr /* OgaModel* */ model,
+                                                                              out IntPtr /* OgaGeneratorParams** */ generatorParams);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern void OgaDestroyGeneratorParams(IntPtr /* OgaGeneratorParams* */ generatorParams);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaGeneratorParamsSetMaxLength(IntPtr /* OgaGeneratorParams* */ generatorParams,
+                                                                                    int maxLength);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaGeneratorParamsSetInputIDs(IntPtr /* OgaGeneratorParams* */ generatorParams,
+                                                                                   int[] /* int32_t* */ inputIDs,
+                                                                                   UIntPtr inputIDsCount,
+                                                                                   UIntPtr sequenceLength,
+                                                                                   UIntPtr batchSize);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaCreateGenerator(IntPtr /* OgaModel* */ model,
+                                                                        IntPtr /* OgaGeneratorParams* */ generatorParams,
+                                                                        out IntPtr /* OgaGenerator** */ generator);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern void OgaDestroyGenerator(IntPtr /* OgaGenerator* */ generator);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern bool OgaGenerator_IsDone(IntPtr /* OgaGenerator* */ generator);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaGenerator_ComputeLogits(IntPtr /* OgaGenerator* */ generator);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaGenerator_GenerateNextToken_Top(IntPtr /* OgaGenerator* */ generator);
+
+        [DllImport(NativeLib.DllName, CallingConvention = CallingConvention.Winapi)]
+        public static extern IntPtr /* OgaResult* */ OgaGenerator_GetSequence(IntPtr /* OgaGenerator* */ generator,
```

Comment:
It would be better to have a separate call to get the size of the pre-allocated buffer. Then we could have int[] in the signature. It would take care of the pinning, now the pinning must be done by C# API.

Reply:
Changed the API to now have two functions. One for getting the count of tokens in the sequence which can be used for allocating the sequence on the c# side and the other for getting the data that can be copied into this buffer.